File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.py
def solution(players, callings):
    player_to_index = {player:i for i, player in enumerate(players)}

    for call in callings:
        cur_idx = player_to_index[call]
        
        prev_idx = cur_idx - 1
        
        players[prev_idx], players[cur_idx] = players[cur_idx], players[prev_idx]

        player_to_index[players[prev_idx]] = prev_idx
        player_to_index[players[cur_idx]] =  cur_idx
        
    return players

File: 프로그래머스/1/178871. 달리기 경주/달리기 경주.java
import java.util.HashMap;
class Solution {
    public String[] solution(String[] players, String[] callings) {
        HashMap<Integer, String> ranks = new HashMap<>();
        HashMap<String, Integer> names = new HashMap<>();
        for (int i = 0; i < players.length; i++) {
            ranks.put(i, players[i]);
            names.put(players[i], i);
        }
        for (int i = 0; i < callings.length; i++) {
            int prevRank = names.get(callings[i]);
            String prevName = ranks.get(prevRank - 1);

            int nextRank = names.get(prevName);
            names.put(prevName, prevRank);
            names.put(callings[i], nextRank);

            ranks.put(prevRank, prevName);
            ranks.put(nextRank, callings[i]);
        }
        return ranks.values().toArray(new String[0]);
    }
}